File: datasets/Python/Gemini/DataStructures/medium5.py
import collections

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def is_bst_helper(node, min_val, max_val):
    if node is None:
        return True

    # Check if the current node's data is within the valid range
    if not (min_val < node.data < max_val):
        return False

    # Recursively check left subtree: update max_val to current node's data
    # Recursively check right subtree: update min_val to current node's data
    return (is_bst_helper(node.left, min_val, node.data) and
            is_bst_helper(node.right, node.data, max_val))

def is_binary_search_tree(root):
    # Initialize with negative and positive infinity for the root's range
    return is_bst_helper(root, float('-inf'), float('inf'))

def build_tree_from_input():
    print("Enter tree nodes in level order (BFS), separated by spaces.")
    print("Use 'null' for missing children (e.g., 3 5 2 1 4 6 null):")
    input_line = input().strip().split()

    if not input_line or input_line[0].lower() == 'null':
        return None

    root_data = int(input_line[0])
    root = Node(root_data)
    queue = collections.deque([root])
    i = 1

    while queue and i < len(input_line):
        current_node = queue.popleft()

        # Left child
        if i < len(input_line) and input_line[i].lower() != 'null':
            left_data = int(input_line[i])
            current_node.left = Node(left_data)
            queue.append(current_node.left)
        i += 1

        # Right child
        if i < len(input_line) and input_line[i].lower() != 'null':
            right_data = int(input_line[i])
            current_node.right = Node(right_data)
            queue.append(current_node.right)
        i += 1
    return root

if __name__ == "__main__":
    root = build_tree_from_input()
    if is_binary_search_tree(root):
        print("Yes")
    else:
        print("No")

File: datasets/C++/Gemini/DataStructures/medium5.cpp
#include <iostream>
#include <queue>
#include <string>
#include <limits>       // Required for numeric_limits
#include <sstream>      // <--- ADD THIS LINE

// Define a Node structure for the binary tree
struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};

// Helper function for BST validation
bool isBSTHelper(Node* node, long long min_val, long long max_val) {
    // An empty tree is a valid BST
    if (node == nullptr) {
        return true;
    }

    // Check if the current node's data is within the valid range
    // Using long long for min_val/max_val to handle edge cases with INT_MIN/INT_MAX data
    if (node->data <= min_val || node->data >= max_val) {
        return false;
    }

    // Recursively check left subtree: update max_val to current node's data
    // Recursively check right subtree: update min_val to current node's data
    return isBSTHelper(node->left, min_val, node->data) &&
           isBSTHelper(node->right, node->data, max_val);
}

// Main function to check if a binary tree is a BST
bool isBinarySearchTree(Node* root) {
    // Initialize with numeric_limits for the root's range
    return isBSTHelper(root, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max());
}

// Function to build a tree from level-order input
Node* buildTreeFromInput() {
    std::cout << "Enter tree nodes in level order (BFS), separated by spaces." << std::endl;
    std::cout << "Use 'null' for missing children (e.g., 3 5 2 1 4 6 null):" << std::endl;
    std::string line;
    std::getline(std::cin, line);

    // This is the line that caused the error, now fixed by adding <sstream>
    std::stringstream ss(line);
    std::string token;

    ss >> token;
    if (token == "null") {
        return nullptr;
    }

    Node* root = new Node(std::stoi(token));
    std::queue<Node*> q;
    q.push(root);

    while (!q.empty() && ss >> token) {
        Node* current_node = q.front();
        q.pop();

        // Left child
        if (token != "null") {
            current_node->left = new Node(std::stoi(token));
            q.push(current_node->left);
        }

        // Right child
        if (ss >> token) { // Try to read next token for right child
            if (token != "null") {
                current_node->right = new Node(std::stoi(token));
                q.push(current_node->right);
            }
        }
    }
    return root;
}

// Function to deallocate tree memory (good practice)
void deleteTree(Node* node) {
    if (node == nullptr) {
        return;
    }
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
}


int main() {
    Node* root = buildTreeFromInput();

    if (isBinarySearchTree(root)) {
        std::cout << "Yes" << std::endl;
    } else {
        std::cout << "No" << std::endl;
    }

    deleteTree(root); // Clean up memory
    return 0;
}